in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1088 @@
+    const abi = this.getEntrypointAbi();
+    const entrypointPackedArgs = PackedValues.fromValues(encodeArguments(abi, [appPayload, setupPayload]));
+
+    return builder
+      .setOrigin(this.address)
+      .setTxContext({
+        chainId: this.chainId,
+        version: this.version,
+        gasSettings: userRequest.gasSettings,
+      })
+      .setFunctionSelector(
+        FunctionSelector.fromNameAndParameters(abi.name, abi.parameters)
+      )
+      .setFirstCallArgsHash(entrypointPackedArgs.hash());
+      .setArgsOfCalls([
+        ...appPayload.packedArguments,
+        ...feePayload.packedArguments,
+         entrypointPackedArgs
+      ])
+      .addAuthWitness(
+        await this.auth.createAuthWit(appPayload.hash())
+      )
+      .addAuthWitness(
+        await this.auth.createAuthWit(setupPayload.hash())
+      )
+  }
+
+  private getEntrypointAbi() {
+    return {
+      name: 'entrypoint',
+      isInitializer: false,
+      // ... same as before
+    }
+  }
+}
+
+```
+
+#### Fill in the `TxExecutionRequest`
+
+The abstract `BaseWallet` can implement:
+
+```ts
+async getTxExecutionRequest(userRequest: UserRequest): Promise<TxExecutionRequest> {
+  if (!userRequest.gasSettings) {
+    throw new Error('Gas settings must be provided');
+  }
+  if (!userRequest.paymentMethod) {
+    throw new Error('Payment method must be provided');
+  }
+
+  const builder = new TxExecutionRequestBuilder();
+
+  for (const call of request.calls) {
+    addMainFunctionCall(builder, call);
+    if (call.deploymentOptions?.registerClass) {
+      addClassRegistration(builder, call);
+    }
+    if (call.deploymentOptions?.publicDeploy) {
+      addPublicDeployment(builder, call);
+    }
+    // if the user is giving us an artifact,
+    // allow the PXE to access it
+    if (call.contractArtifact) {
+      builder.addTransientContract({
+        artifact: call.contractArtifact,
+        instance: call.contractInstance,
+      });
+    }
+  }
+
+  // Add stuff needed for setup, e.g. function calls, auth witnesses, etc.
+  await userRequest.paymentMethod.adaptTxExecutionRequest(builder, userRequest);
+
+  // Adapt the request to the entrypoint in use.
+  // Since BaseWallet is abstract, this will be implemented by the concrete class.
+  this.adaptTxExecutionRequest(builder, userRequest);
+
+  return builder.build();
+
+}
+```
+
+### BaseWallet.#simulateInner
+
+```ts
+// Used by simulate and read
+async #simulateInner(userRequest: UserRequest): ReturnType<BaseWallet['simulate']> {
+  const txExecutionRequest = await this.getTxExecutionRequest(userRequest);
+  const simulatedTx = await this.simulateTx(txExecutionRequest, builder.simulatePublicFunctions, builder.from); 
+  const decodedReturn = decodeSimulatedTx(simulatedTx, builder.functionAbi);
+  return {
+    tx: simulatedTx.tx,
+    publicOutput: simulatedTx.publicOutput,
+    privateOutput: simulatedTx.privateReturnValues,
+    executionResult: simulatedTx.executionResult,
```

Comment:
`tx` is our transaction object (PrivateKernelTailCircuitPublicInputs, private proof, logs, enqueued public functions)

`publicOutput` is 
```ts
/**
 * Outputs of processing the public component of a transaction.
 */
export class PublicSimulationOutput {
  constructor(
    public encryptedLogs: EncryptedTxL2Logs,
    public unencryptedLogs: UnencryptedTxL2Logs,
    public revertReason: SimulationError | undefined,
    public constants: CombinedConstantData,
    public end: CombinedAccumulatedData,
    public publicReturnValues: NestedProcessReturnValues[],
    public gasUsed: Partial<Record<PublicKernelType, Gas>>,
  ) {}
```

The `privateReturnValues` are:
```ts
/** Return values of simulating a circuit. */
export type ProcessReturnValues = Fr[] | undefined;

/** Return values of simulating complete callstack. */
export class NestedProcessReturnValues {
  values: ProcessReturnValues;
  nested: NestedProcessReturnValues[];
```

and `executionResult` is a bit redundant in that it includes all the stuff in `tx`, but also stuff we need for proving like the verification key. It may be that that is all we truly need from this.